zerver/actions/message_edit.py

```diff
@@ -918,7 +918,7 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
             else:
                 # This corresponds to the case when messages are moved
                 # to a stream-topic pair that didn't exist.
-                if new_visibility_policy == UserTopic.VisibilityPolicy.INHERIT:
+                if new_visibility_policy == UserTopic.VisibilityPolicy.INHERIT:  # nocoverage
```

Comment:
I have tried to explain this in the commit message, but would mention here at weill with code.

There will be no UserTopic rows for original topic with visibility set to `INHERIT` since we delete them.

And since the target topic has no messages, there will be no UserTopic rows for it. So, the code below also does not add any entry to `orig_topic_user_profile_to_visibility_policy`.
```python
user_profiles_having_visibility_policy: Set[UserProfile] = set(
    itertools.chain(
        orig_topic_user_profile_to_visibility_policy.keys(),
         target_topic_user_profile_to_visibility_policy.keys(),
     )
)

user_profiles_for_visibility_policy_pair: Dict[
    Tuple[int, int], List[UserProfile]
] = defaultdict(list)
for user_profile_with_policy in user_profiles_having_visibility_policy:
    if user_profile_with_policy not in target_topic_user_profile_to_visibility_policy:
        target_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT
    elif user_profile not in orig_topic_user_profile_to_visibility_policy:
        orig_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT

        orig_topic_visibility_policy = orig_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ]
        target_topic_visibility_policy = target_topic_user_profile_to_visibility_policy[
             user_profile_with_policy
        ]
        user_profiles_for_visibility_policy_pair[
             (orig_topic_visibility_policy, target_topic_visibility_policy)
        ].append(user_profile_with_policy)
``` 

So, there is actually no entry in `user_profiles_for_visibility_policy_pair` with `orig_topic_visibility_policy` set to `INHERIT`.

Reply:
I am not sure whether we want to remove the code itself or just keep it for safety purposes to avoid unnecessary operations.

Reply:
@prakhar1144 FYI. Can you confirm if I am missing something here?

Reply:
> adds the entry in user_profiles_for_visibility_policy_pair with orig_topic_visibility_policy set to INHERIT.
Ok, I got this. But the below code will run only if the user_profile exists in `user_profiles_having_visibility_policy` and does not exist in `orig_topic_user_profile_to_visibility_policy`. So, it needs to exist in `target_topic_user_profile_to_visibility_policy`, but it cannot since there are no messages in the target topic (essentially meaning no topic exists) and hence there should not be any `UserTopic` rows for it, correct?

```python
elif user_profile not in orig_topic_user_profile_to_visibility_policy:
       orig_topic_user_profile_to_visibility_policy[
           user_profile_with_policy
       ] = UserTopic.VisibilityPolicy.INHERIT
```

Reply:
In that case, we are dealing with the set of users
target_topic: INHERIT (No UserTopic row)
original_topic: INHERIT (No UserTopic row)

so, user_profile neither exists in `orig_topic_user_profile_to_visibility_policy` nor `target_topic_user_profile_to_visibility_policy` hence not in `user_profiles_having_visibility_policy`.


And we don't need to handle the "move user topic records" in this case, simply move the message. For those users, target_topic visibility policy is already INHERIT (i.e no UserTopic row exists).

-----
Also, is CZO down? -- can you please confirm. It just keeps loading for me.

Reply:
Yeah so, we cannot reach this if-conditon then.

> Also, is CZO down? -- can you please confirm. It just keeps loading for me.

yes same for me.

Reply:
> Yeah so, we cannot reach this if-conditon then.

if
target_topic has no messages (hence, no UserTopic row)
orig_topic: MUTED

Now,
`orig_topic_user_profile_to_visibility_policy` has user_profile
`user_profiles_having_visibility_policy` has user_profile
```python
if user_profile_with_policy not in target_topic_user_profile_to_visibility_policy:
        target_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT
```

Now, the if block will be reached, right?

Reply:
Ah, right. -- that can be removed.

Reply:
I think it's possible to have a `UserTopic` row for a topic with no messages. I think the case for this code path is a message is sent to a topic, and then you mute the topic, and then someone deletes the message, and then someone tries to move messages from a topic that you've unmuted in some other stream into the (empty of messages, but muted) topic name.

Reply:
Hm, I can [recall a discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/duplicate.20request.20to.20update.20UserTopic.20row/near/1523816) around this on CZO.

> So in short there should be no requirement that the topic already have messages in it when creating a UserTopic record.

Reply:
Hmm, I agree there can be such case. But in such cases as well, the current code seems wrong. Because we just skip the case when the `new_visibility_policy` is `INHERIT`, and I think we do so assuming there is no `UserTopic` row for a topic with no messages. 

I have updated the code and added the tests too. Will push to this PR soon.